src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -700,4 +706,143 @@ static VaList empty() {
+            LONGDOUBLE(findBSM("C_LONGDOUBLE"), false),
+            POINTER(findBSM("C_POINTER"), false);
+
+            private final DynamicConstantDesc<ValueLayout> bsm;
+            private final boolean isIntegral;
+
+            Kind(DynamicConstantDesc<ValueLayout> bsm, boolean isIntegral) {
+                this.bsm = bsm;
+                this.isIntegral = isIntegral;
+            }
+
+            public DynamicConstantDesc<ValueLayout> bsm() {
+                return bsm;
+            }
+
+            public boolean isIntergral() {
+                return isIntegral;
+            }
+
+            public boolean isPointer() {
+                return this == POINTER;
+            }
+
+            private static DynamicConstantDesc<ValueLayout> findBSM(String fieldName) {
+                return DynamicConstantDesc.ofNamed(
+                    BSM_GET_STATIC_FINAL,
+                    fieldName,
+                    CValueLayout.class.describeConstable().orElseThrow(),
+                    CLinker.class.describeConstable().orElseThrow()
+                );
+            }
+        }
+
+        private final Kind kind;
+
+        protected CValueLayout(Kind kind, ByteOrder order, long size, long alignment, Map<String, Constable> attributes) {
+            super(order, size, alignment, attributes);
+            this.kind = kind;
+        }
+
+        public static CValueLayout ofChar(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.CHAR, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofShort(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.SHORT, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofInt(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.INT, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofLong(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.LONG, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofLongLong(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.LONGLONG, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofFloat(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.FLOAT, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofDouble(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.DOUBLE, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofLongDouble(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.LONGDOUBLE, order, bitSize, bitSize, Map.of());
+        }
+
+        public static CValueLayout ofPointer(ByteOrder order, long bitSize) {
+            return new CValueLayout(Kind.POINTER, order, bitSize, bitSize, Map.of());
+        }
+
+        public final Kind kind() {
```

Comment:
Same here - should this exposed?

Reply:
Yes, kind should be exposed. The classification logic needs to access it, and if we don't expose it it's not possible to extend this class, since the kind is a constructor argument and we need to call the constructor.

Reply:
I guess the fact that classification logic needs it can be handle by reorganizing the code so that the impl private.
But your concern about subclasses needing to pass a kind is true.